sql-plugin/src/main/scala/org/apache/spark/sql/catalyst/json/rapids/GpuJsonScan.scala

```diff
@@ -0,0 +1,326 @@
+    throw new IllegalStateException("ROW BASED PARSING IS NOT SUPPORTED ON THE GPU...")
+  }
+
+  override def buildColumnarReader(partFile: PartitionedFile): PartitionReader[ColumnarBatch] = {
+    val conf = broadcastedConf.value.value
+    val reader = new PartitionReaderWithBytesRead(new JsonPartitionReader(conf, partFile,
+      dataSchema, readDataSchema, parsedOptions, maxReaderBatchSizeRows, maxReaderBatchSizeBytes,
+      metrics))
+    ColumnarPartitionReaderWithPartitionValues.newReader(partFile, reader, partitionSchema)
+  }
+}
+
+class JsonPartitionReader(
+    conf: Configuration,
+    partFile: PartitionedFile,
+    dataSchema: StructType,
+    readDataSchema: StructType,
+    parsedOptions: JSONOptions,
+    maxRowsPerChunk: Integer,
+    maxBytesPerChunk: Long,
+    execMetrics: Map[String, GpuMetric])
+  extends GpuTextBasedPartitionReader(conf, partFile, dataSchema, readDataSchema,
+    parsedOptions.lineSeparatorInRead, maxRowsPerChunk, maxBytesPerChunk, execMetrics) {
+
+  def buildJsonOptions(parsedOptions: JSONOptions): cudf.JSONOptions = {
+    val builder = cudf.JSONOptions.builder()
+    builder.build
+  }
+
+  /**
+   * Read the host buffer to GPU table
+   *
+   * @param dataBuffer     host buffer to be read
+   * @param dataSize       the size of host buffer
+   * @param cudfSchema     the cudf schema of the data
+   * @param readDataSchema the Spark schema describing what will be read
+   * @param hasHeader      if it has header
+   * @return table
+   */
+  override def readToTable(
+      dataBuffer: HostMemoryBuffer,
+      dataSize: Long,
+      cudfSchema: Schema,
+      readDataSchema: StructType,
+      hasHeader: Boolean): Table = {
+
+    val jsonOpts = buildJsonOptions(parsedOptions)
+    Table.readJSON(cudfSchema, jsonOpts, dataBuffer, 0, dataSize)
+  }
+
+  /**
+   * File format short name used for logging and other things to uniquely identity
+   * which file format is being used.
+   *
+   * @return the file format short name
+   */
+  override def getFileFormatShortName: String = "JSON"
+
+  /**
+   * Handle the table decoded by GPU
+   *
+   * @param readDataSchema the Spark schema describing what will be read
+   * @param table          the table decoded by GPU
+   * @return the new optional Table
+   */
+  override def handleResult(readDataSchema: StructType, table: Table): Option[Table] = {
+    val tableCols = table.getNumberOfColumns
+    if (readDataSchema.length == tableCols) {
+      return Some(table)
+    }
+
+    // JSON will read all columns in dataSchema due to https://github.com/rapidsai/cudf/issues/9990,
+    // but actually only the columns in readDataSchema are needed, we need to do the "column" prune,
+    // Once the FEA is shipped in CUDF, we should remove this hack.
+
+    val prunedCols = readDataSchema.fieldNames.map { name =>
+      val optionIndex = dataSchema.getFieldIndex(name)
+      closeOnExcept(table) { _ =>
```

Comment:
Same "who really owns the table?" comments apply here as well.

Reply:
same as above reply